src/client/common/utils/decorators.ts

```diff
@@ -266,3 +266,36 @@ export function testOnlyMethod() {
         return descriptor;
     };
 }
+
+// Mark a method that returns a promise to chain it
+export function chainable() {
```

Comment:
I know that you had looked at this before, so I don't want to rehash a debate on this. But I'm not keen on something that changes the way the function works this much being "hidden" in a decorator. I like decorators for stuff like logging or error handling. Something like this feels like a big functionality difference which I wouldn't want changed by a decorator. They can also be a pain to debug at times.

I guess if this PR looks good and the chaining cases are all handed then I won't hold things up. But if we run into issue in the future with this decorator hiding real issues or making debugging hard I'd argue against it.

Reply:
Debugging it isn't hard, you just set breakpoints in the decorator. The reason for it is so we don't have duplicate code for chaining a promise. Not sure how else to implement common code for promise chaining though. A wrapper method instead?

Something like:
```
public doStuffWithPromise() {
   return chained(async () => {
       await doingstuff();
   })
}
```

Does that grok better? It's essentially the same thing as the decorator but more explicit.

Reply:
Probably just anti decorator bias on my part. I read decorators like this:

```
@thisStuffDoesn'tReallyMatter
@youCanReadTheFunctionJustFineWithoutThis
@probablyJustLoggingOrSomethingLikeThat
function doStuff() {}
```
This feels more like:
```
@thisIsReallyKindaABigChangeToHowThisFunctionWorks
function doStuff() {}
```

But there is no reason that a decorator can't do something like you are doing here. So I vote we go for it. I was thinking it might cause issues with the case where there is a python error in the chain (where we return a false and show a message) but it looks like you added that in a later commit.